---
output: github_document
#output: html_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# GraphR overview

<center><img src="GraphR_Method_plot_final.png" width="600" height="600" ></center>

The GraphR (Graphical Regression) is a flexible approach which incorporates sample heterogenity and enables covariate-dependent graphs. Our regression-based method provides a functional mapping from the covariate space to precision matrix for different types of heterogeneous graphical model settings. GraphR imposes sparsity in both edge and covariate selection and computationally efficient via use of variational Bayes algorithms. The method is versatile to incorporate different type of covariates such as <br>
(I) **binary** (control and disease specific graphs), <br>
(II) **categorical** (category specific graphs such as cancer subtypes), <br>
(III) **univariate continuous** (time varying graphs for single cell data), <br>
(IV) **categorical + univariate continuous** (graphs changing over category such as cancer sub-types and continuous scale as biomarkers), <br> 
(V) **multivariate continuous** (spatial transcriptomics co-expression networks). <br>
GraphR is implemented as an open-source R package and Shiny app.


## Installation

You can install the released version of GraphR from (https://github.com/bayesrx/GraphR) with:

``` {r , eval=FALSE}
devtools::install_github("bayesrx/GraphR")
library(GraphR)
```

## Functions

### GraphR_est() function 

The **GraphR_est()** function can be used to estimate the graphical regression coefficients and inclusion probabilities of external covariates for the GraphR models. It is suggested to maintain $n/pq >1$ and efficacy of the method increase with high values of $n/pq$ ratio. For priors, we assume $\pi \sim Beta(a_\pi, b_\pi)$ and $\tau \sim \Gamma(a_\tau, b_\tau)$.

The **inputs** of estimation function are given below.

+ **Features (nodes)**: Nodes of the graphs among which edges are built (e.g. a gene expression matrix of dimensions $n \times p$). **Please standardize features and external covariates before plugging into the function**. 

+ **External (external covariates)**: A $n \times q$ matrix of external covariates. **Please standardize continuous external covariates before plug into the estimation function.**

+ **$\boldsymbol a_{\boldsymbol \pi}$, $\boldsymbol b_{\boldsymbol \pi}$**: Hyper-parameters from $\pi$ $\sim$ Beta$(a_\pi, b_\pi)$.

+ **$\boldsymbol a_{\boldsymbol \tau}$, $\boldsymbol b_{\boldsymbol \tau}$**: Hyper-parameters from $\tau$ $\sim$ Gamma($a_{\tau}$, $b_{\tau}$).

+ **Max\_iter**: Maximum number of iterations.

+ **Max\_tol**: Maximum tolerance.

**Outputs** of the **GraphR_est()** function are provided below. 

+ **Beta (the graphical regression coefficients)**: A $p \times p \times q$ array of coefficients for external covariates. The $[i,j,k]$ element represents the effect of k-th external covariates on regression of j-th node on i-th node.

+ **Phi (posterior inclusion probability)**: A $p \times p \times q$ array storing posterior inclusion probability (PIP) of external covariates. The \eqn{[i,j,k]} elements represents the PIP of k-th external covariates on regression of j-th node on i-th node.

+ **Omega_diag (diagonal elements of precision matrix)**: A p vector with i-th element representing the inverse variance of error.


### GraphR_pred() function
The **GraphR_pred()** function can be used to predict partial correlation between two nodes and the corresponding inclusion probabilities from the results of GraphR model alongwith Bayesian FDR-adjusted p-values.

The **inputs** of estimation function are given below.

+ **New_df**: A matrix of new external covarites based on which predicitons are made. **Note: Please standardize continuous external covariates before plug into the estimation function.**

+ **GraphR_est_res**: Results from `GraphR_est` function.

+ **Beta**: A $p \times p \times q$ array storing coefficients of external covariates. The $[i,j,k]$ elements represents the effect of k-th external covariates on regression of j-th node on i-th node.

+ **Omega_diag**: A p vector with i-th element representing the inverse variance of error.

+ **Pip**: A $p \times p \times q$ array storing posterior inclusion probability (PIP) of external covariates. The $[i,j,k]$ elements represents the PIP of k-thcexternal covariates on regression of j-th node on i-th node.


The **output** contains following information. 

+ **Feature_id1**, **feature_id2**: Indices of features or nodes.

+ **Pr_inclusion**: Posterior inclusion probability of connections between two nodes
based on "And" rules.

+ **Correlation**: Partial correlation between two nodes. Values with maximum magnitudes are provided.

+ **FDR_p**: Bayesian FDR-adjusted p values.


## Example 
An example code with one of the existing datasets to demonstrate how to run the functions and obtain inference.


```{r}
set.seed(100)
library(GraphR)
library(dplyr)
data("Pam50")

features <- apply(Pam50$features,2,scale) %>% as.matrix()
features[c(1:5),c(1:5)]

external <- Pam50$external %>% as.matrix()
external[c(1:5),]


system.time(res <- GraphR_est(
  features,
  external,
  a_pi = 1,
  b_pi = 4,
  a_tau = 0.005,
  b_tau = 0.005,
  max_iter = 2000,
  max_tol = 0.001
))

####### prediction
new_df <- diag(3)
colnames(new_df) <- colnames(external)

system.time(pred <- GraphR_pred(new_df, res))
head(pred)
```


## Shiny App 
(link to the Shiny App tab only)

## Tutorial website
(link to the tutorial website)

## Paper 
(Link to the paper) 

## Supplementary file 
(link to the bookdown file)
